src/main/kotlin/org/rust/lang/core/resolve/scope/RustResolveScope.kt

```diff
@@ -1,9 +1,256 @@
 package org.rust.lang.core.resolve.scope
 
-import org.rust.lang.core.psi.RustCompositeElement
-import org.rust.lang.core.psi.RustNamedElement
+import com.intellij.openapi.module.Module
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.psi.util.PsiTreeUtil
+import com.intellij.psi.util.PsiUtilCore
+import org.rust.cargo.util.AutoInjectedCrates
+import org.rust.cargo.util.cargoProject
+import org.rust.cargo.util.getPsiFor
+import org.rust.cargo.util.preludeModule
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.impl.RustFile
+import org.rust.lang.core.psi.impl.mixin.isStarImport
+import org.rust.lang.core.psi.impl.rustMod
+import org.rust.lang.core.psi.util.module
 
 interface RustResolveScope : RustCompositeElement {
-    val declarations: Collection<RustNamedElement>
+    data class Context(
```

Comment:
>No, both of those two are integral parts of the resolving procedure (and not the marking interface RustResolveScope) which should be encapsulated (or at least confined) by the single 'module' -- this module is historically RustResolveEngine.

Yes, this is part of the resolve, and that's why it's in the org.rust.lang.core.**resolve**.scope package. `RustResolveEngine` does a bunch of other stuff, besides scopes, so by extracting scoping rules into a 'submodule' we can separate concerns. And this is possible because the dependency between scopes and RRE is unidirectional.

Note that another consumer of scopes API is the [completion engine], and this PR actually fixes #405 because the same scoping logic is now used for both resolve and completion. Hm, interesting, it also fixes completion for the `*`-imported items, will need to add a test for it. 

[completion engine]: https://github.com/intellij-rust/intellij-rust/pull/444/files#diff-f6a8fb5e98ccbac0c27f3afcc887a52eR43